core/local/steps/win_detect_move.js

```diff
@@ -0,0 +1,101 @@
+/* @flow */
+
+const { id } = require('../../metadata')
+const Buffer = require('./buffer')
+const logger = require('../../logger')
+const log = logger({
+  component: 'winDetectMove'
+})
+
+// Wait at most this delay (in milliseconds) to see if it's a move.
+const DELAY = 1000
+
+/*::
+import type { AtomWatcherEvent } from './event'
+import type Pouch from '../../pouch'
+
+type PendingBatch = {
+  events: AtomWatcherEvent[],
+  deleted: Map<string, string>,
+  timeout: TimeoutID
+}
+*/
+
+// TODO add unit tests and logs
+// TODO check that a file/dir created and removed just after is not seen as a move
+
+// On windows, ReadDirectoryChangesW emits a deleted and an added events when
+// a file or directory is moved. This step merges the two events to a single
+// renamed event.
+async function winDetectMove (buffer, out, pouch) {
+  const pending /*: PendingBatch[] */ = []
+  const sendReadyBatches = () => {
+    while (pending.length > 0) {
+      if (pending[0].deleted.size !== 0) {
+        break
+      }
+      const p = pending.shift()
+      clearTimeout(p.timeout)
+      out.push(p.events)
+    }
+  }
+
+  while (true) {
+    // Wait for a new batch of events
+    const events = await buffer.pop()
+
+    // First, push the new events in the pending queue
+    const deleted = new Map()
+    for (const event of events) {
+      if (event.action === 'deleted') {
+        const release = await pouch.lock('winMoveDetector')
+        try {
+          const was = await pouch.db.get(id(event.path))
+          deleted.set(was.fileid, event.path)
+        } catch (err) {
+          // Ignore the error
+        } finally {
+          release()
+        }
+      }
+    }
+    const timeout = setTimeout(() => {
+      out.push(pending.shift().events)
+      sendReadyBatches()
+    }, DELAY)
+    pending.push({ events, deleted, timeout })
+
+    // Then, see if a created event matches a deleted event
+    for (const event of events) {
+      if (event.action === 'created') {
+        for (let i = 0; i < pending.length; i++) {
+          const path = pending[i].deleted.get(event.stats.fileid)
+          if (!path || path === event.path) {
+            continue
+          }
+          const l = pending[i].events.length
+          for (let j = 0; j < l; j++) {
```

Comment:
Can we have multiple `deleted` events for a unique `created` event when renaming a file/folder?
If not, aren't `pending[i].deleted.get(fileId)` and `pending[i].events[j]` (which has the same file id) the same object? In this case we could avoid this loop entirely.

Reply:
Or maybe it's necessary for the `splice` just a few lines below.
But I'd rather have the loop used for this purpose only for the sake of clarity.

Reply:
> Can we have multiple deleted events for a unique created event when renaming a file/folder?
No, I don't we can that.

> If not, aren't pending[i].deleted.get(fileId) and pending[i].events[j] (which has the same file id) the same object?
No, the first one is a string (the path), the other is an object (the event).

> In this case we could avoid this loop entirely.
We need the loop for `splice`.

> But I'd rather have the loop used for this purpose only for the sake of clarity.
Why not, but I'm not sure to see what you want to move and where. Can you make a suggestion?

Reply:
No need for suggestions. Avoiding the loop would have been possible if `pending[i].deleted.get(fileId)` and `pending[i].events[j]` would both have been events. In this case I would have prefered to use the getter to access the event since it clearly shows we can only have one `deleted` event for any single `created` event.